src/main/java/software/amazon/cloudwatchlogs/emf/logger/MetricsLogger.java

```diff
@@ -157,8 +185,13 @@ public MetricsLogger putMetric(String key, double value) {
      * @return the current logger
      */
     public MetricsLogger putMetadata(String key, Object value) {
-        this.context.putMetadata(key, value);
-        return this;
+        rwl.readLock().lock();
+        try {
+            this.context.putMetadata(key, value);
+            return this;
+        } finally {
+            rwl.readLock().unlock();
+        }
     }
```

Comment:
One cleaner way to avoid multiple readlock lock and unlock across multiple putXX method is

```
public MetricsLogger putMetric(String key, double value) {
        return acquireLockForPut(() -> {
            this.context.putMetric(key, value, Unit.NONE);
            return this;
        });
    }

    private MetricsLogger acquireLockForPut(Supplier<MetricsLogger> any) {
        rwl.readLock().lock();
        try {
            return any.get();
        } finally {
            rwl.readLock().unlock();
        }
    }
```

Reply:
Wow that's cool...I'll take look into it.